AVM2 Prototype De-Emphasis & VM Correctness

kmeisthax opened the pull request:
AS3 uses the ES4 class system, a completely separate object model to the prototype-based ES3 one. While AVM2 supports both models, ES4 classes only retain prototypes as a compatibility mechanism. Despite this, our AVM2 implementation extensively abuses prototypes all over the place. This PR attempts to remove as many inappropriate uses of prototypes as possible - everything should hit the class constructor or class itself; with prototypes *only* being a property bag for legacy lookups.

 * [x] Split `ClassObject` off from `FunctionObject`.
 * [x] Create a `constr` chain of base class constructor objects to mirror the `proto` chain on instances.
 * [x] Move `nnn_deriver` functions into `Class`. This allows classes to specify the native object representation of themselves and all of their subclasses.
 * [x] Associate instances with a `constr` as well as a `proto`, *dis*associate prototypes from any `class` or `constr`
 * [x] Change trait lookup to use the `constr` chain instead of the `proto` chain
 * [x] Change `TObject.construct` into a function to be called on constructors (rather than prototypes). `construct` now handles the entire object construction process for both ES3 (`FunctionObject`) and ES4 (`ClassObject`) constructors, which run in different ways. Calling `construct` on any other function type yields an error.
 * [x] Remove references to `system_prototypes`, and replace with `system_constructors`.
 * [x] Change ES4 type checks to use constructors where possible.
 * [x] Add a notion of "native class initializers" separate from the existing initializer chain. This will allow non-constructable classes to still be initialized via `super`.
 * [x] Initialize objects in native allocation functions where possible. e.g. `ByteArrayObject::new` should initialize the returned object. (This is almost certainly *not* possible for `StageObject`...)
 * [x] Make native allocation functions pull the prototype out of the constructor instead of asking for one, as it is almost always the case that we want `Class.prototype` to be the prototype here.
 * [x] Ensure all class initializers run, including native ones.
 * [x] Remove lazy trait instantiation and instead install all traits at allocation time.
 * [x] Separate trait instantiation from class/function initialization so we don't do that twice
 * [x] Enforce `sealed`, `final`, `interface` on classes
 * [x] Enforce `final` and `override` on traits
 * [x] Enforce parameter types on user methods
 * [x] Implement implicit coercions on parameter typechecks
 * [x] Tie `op_coerce` into the same implicit coercions that parameter typechecks do
 * [x] Make `is` yield true for coerceable types
 * [x] Enforce parameter types on builtins
 * ~Add parameter types to all builtins~ Postponed as it interferes with already-finished work to store builtin methods and properties in static tables
 * ~AS3 builtin display object subclasses (e.g. `MovieClip`, `Shape` etc) should register themselves as constructors for empty symbols in their class init instead of using the instance initializer to do the same thing.~ Tried this, it started spitting out all sorts of weird errors in AVM1 for some reason.

kmeisthax commented:
I believe this to be ready for review. The few things I wasn't able to accomplish are minor; and this PR removes a significant number of unnecessary workarounds and hacks we were carrying forward on AVM2. The only PR it imperils is Vector, but that will probably benefit from the cleanup anyway.

Herschel commented on core/src/avm2/activation.rs:
```diff
@@ -1203,22 +1306,23 @@ impl<'a, 'gc, 'gc_context> Activation<'a, 'gc, 'gc_context> {
+            });
+        let base_constr = base_constr?;
+
+        let value = base_constr.call_instance_getter(&name, Some(object), self)?;
```
This is incorrectly assuming that the super property is a getter (ditto with `setsuper`). Consider:
```
class Base {
	var a: int = 1;
}
class Sub extends Base {
	function f() { return super.a; }
	function g() { super.a++; }
}
var sub = new Sub();
trace(sub.f());
sub.g();
trace(sub.f());
```

Herschel commented on core/src/avm2/class.rs:
```diff
@@ -28,6 +32,82 @@ bitflags! {
+    let mut base_class = constr.as_class();
+    let mut instance_deriver = None;
+
+    while let (Some(b_constr), Some(b_class)) = (base_constr, base_class) {
+        let base_deriver = b_class.read().instance_deriver();
+
+        if base_deriver as usize != implicit_deriver as usize {
+            instance_deriver = Some(base_deriver);
+            break;
+        }
+
+        base_constr = b_constr.base_class_constr();
+        base_class = base_constr.and_then(|c| c.as_class());
+    }
```
Instead of doing this at object construction time, I wonder if this be done once when a class is created, and cache the specific deriver for later.

kmeisthax replied:
It's a little too early to do it in `Class`, since that only holds static information about a class (e.g. names, traits, and so on). At the point that `Class` gets constructed all it gets is the multiname to look up it's superclass with.

However, caching this in `ClassObject` would work.

Herschel reviewed (changes requested):
Thanks for your work! Sorry for the long list of notes as it takes some time to review.

Mainly I am not a fan of the phrase "class constructor" -- I feel that nearly every instance of `constructor` or `constr` in these changes should be renamed to `class`.

I assume the motivation is that you construct an object both via a class and the old ES way via `new someFunction()`. But constructing via function is basically unheard of in AS3 code. The constructor in practice is always a ClassObject, so the code should read as such. Traits only exist on objects created via classes, so all of the methods relating to traits can simply mention `class` or `trait` and it will be clearer. Even the ES built-ins which have the dual traits/prototype setup are still class objects. The code will be clearer with this rename, and easier on the eyes and fingers.

Perhaps as traits improve, some of this will morph into a `Traits` object, but I would definitely like to avoid overusing `constructor`.

I also agree that `Deriver` is somewhat confusing. Maybe `Allocator` or `Instantiator` is better because its job is to allocate a specific object type. (This could be a separate PR since this code was already in place).

Almost every `native_instance_init` is simply calling super. Can we make this a common function somewhere and reuse this?

kmeisthax commented on core/src/avm2/activation.rs:
```diff
@@ -180,31 +181,121 @@ impl<'a, 'gc, 'gc_context> Activation<'a, 'gc, 'gc_context> {
+            )?);
+        }
+
+        match user_arguments.len().cmp(&signature.len()) {
```
I originally did that but Clippy flagged it. Should I `#[allow]` it instead?

Herschel replied:
Guess this is one area where clippy and I disagree. :-) Let's leave it then.

kmeisthax commented on core/src/avm2/object/stage_object.rs:
```diff
@@ -29,40 +46,83 @@ pub struct StageObjectData<'gc> {
+
+        constr.call_native_initializer(Some(this.into()), &[], activation, Some(constr))?;
+
+        Ok(this)
```
We need `graphics` to know *which* `DisplayObject` it's writing to, so we need the extra slot that `StageObject` has.

The only weirdness I think this enables is that you might be able to `mc.addChild(mc2.graphics)`, which would be equivalent to `mc.addChild(mc2)`. This is purely because our `addChild` doesn't have param metadata to enable the typecheck for `DisplayObject`.

Alternatively if this is too weird, we could have `Graphics` store it's associated display object on itself in the Ruffle namespace. It'd add an extra property lookup to graphics operations which is why I didn't want to do that.

Herschel replied:
Maybe in the future if we have a slightly less-annoying way to do inheritance, we could make a separate `GraphicsObject` type that can contain its own pointer.

kmeisthax commented on core/src/avm2/object.rs:
```diff
@@ -727,49 +615,217 @@ pub trait TObject<'gc>: 'gc + Collect + Debug + Into<Object<'gc>> + Clone + Copy
-    /// Construct a host object of some kind and return its cell.
+    /// Call the object, but yield an error if the function is being called
+    /// with too many parameters.
+    fn call_strict(
+        self,
+        _reciever: Option<Object<'gc>>,
+        _arguments: &[Value<'gc>],
+        _activation: &mut Activation<'_, 'gc, '_>,
+        _base_constr: Option<Object<'gc>>,
+    ) -> Result<Value<'gc>, Error> {
+        Err("Object is not strict callable".into())
+    }
```
So, the problem with this is that we have a handful of `Array` tests that call methods like `filter`, `sort`, etc with methods whose signatures are incompatible with what `Array` calls it with, and there doesn't appear to be any "this is an unchecked function" bit on the method signature itself.

The only other way I could think of to make those tests work would be to make `newfunction` flag it's returned function as unchecked. Does this sound correct to you?

Herschel replied:
Yea, this sounds good to me; the callbacks passed to `Array.filter` etc. can be either checked/unchecked depending on their signature. So functions can hold a flag whether they should be checked/unchecked. 

Here it is in avmplus:
https://github.com/adobe/avmplus/blob/master/core/MethodInfo.cpp#L642-L664

Herschel commented on core/src/avm2/activation.rs:
```diff
@@ -1203,22 +1303,23 @@ impl<'a, 'gc, 'gc_context> Activation<'a, 'gc, 'gc_context> {
+            });
+        let superclass_object = superclass_object?;
+
+        let value = superclass_object.call_instance_getter(&name, Some(object), self)?;
```
This is still incorrectly requiring that the property is a getter, and `setsuper` requiring that the property is a setter. Try:

```
class Foo {
	public var a:int = 2;
}

class Bar extends Foo {
	function f() {
		trace(a);
		trace(super.a);
		super.a = 10;
		trace(a);
		trace(super.a);
	}
}

var bar:Bar = new Bar();
bar.f();
```

Herschel reviewed (changes requested):
Thanks! The only other thing I'd like to be in place before merging, if possible, is only removing `call_strict` in favor of `call` that is always strict except for unchecked fns.

kmeisthax commented:
Regarding supercalls: I've done some tests and it appears that the supercall/superget/superset ops fall back to ordinary behaviors on the `receiver`, but only if the property is non-dynamic. So I've implemented these functions this way and renamed them appropriately.

kmeisthax commented:
Alright. Everything you flagged in the last two reviews should be done and this should be ready for (re)review/merge again.

adrian17 commented on core/src/avm2/object.rs:
```diff
@@ -89,12 +94,6 @@ pub trait TObject<'gc>: 'gc + Collect + Debug + Into<Object<'gc>> + Clone + Copy
         name: &QName<'gc>,
         activation: &mut Activation<'_, 'gc, '_>,
     ) -> Result<Value<'gc>, Error> {
-        if !self.has_instantiated_property(name) {
```
now that this is removed, `&mut self` is no longer needed, right? Then `mut` removal can propagate into a lot of places, like `resolve()`, and then every place that does `scope.write(gc).resolve(...)` -> `scope.read().resolve(...)` etc.

Herschel commented:
LGTM! Let's go ahead and get this in so that it's not a blocker. Thanks for the patience!
